src/Shared/TypeLoader.cs

```diff
@@ -169,8 +168,33 @@ private static Assembly LoadAssembly(AssemblyLoadInfo assemblyLoadInfo)
                 // NOTE: don't use ErrorUtilities.VerifyThrowFileExists() here because that will hit the disk again
                 throw new FileNotFoundException(null, assemblyLoadInfo.AssemblyLocation, e);
             }
+        }
+
+        private static Assembly LoadAssemblyUsingMetadataLoadContext(AssemblyLoadInfo assemblyLoadInfo)
+        {
+            string path = assemblyLoadInfo.AssemblyFile;
+
+            if (path is null)
+            {
+#if NETFRAMEWORK
+                AppDomainSetup setup = AppDomain.CurrentDomain.SetupInformation;
+                setup.LoaderOptimization = LoaderOptimization.SingleDomain;
+                AppDomain appDomain = AppDomain.CreateDomain("appDomainToFindPath", null, setup);
+                path = appDomain.Load(new AssemblyName(assemblyLoadInfo.AssemblyName)).Location;
+                AppDomain.Unload(appDomain);
+#else
+                AssemblyLoadContext alc = new("loadContextToFindPath", true);
+                path = alc.LoadFromAssemblyName(new AssemblyName(assemblyLoadInfo.AssemblyName)).Location;
+                alc.Unload();
+#endif
+            }
+
+            string[] runtimePaths = Directory.GetFiles(RuntimeEnvironment.GetRuntimeDirectory(), "*.dll");
```

Comment:
Can this be cached in a static so we don't have to grovel the disk for every UsingTask?

Reply:
I tried doing this and using Concat as you suggested below, and MetadataLoadContext failed to load the assembly. Not even that the assembly was missing information; it just insta-failed. I have no idea why, since it seemed eminently reasonable.

Reply:
Should we debug that? Explictly enumerating the same directory for every task (once we move to detecting taskhostness) seems likely to be expensive.